test/testfloat/testfloat.cpp

```diff
@@ -0,0 +1,395 @@
+// Fizzy: A fast WebAssembly interpreter
+// Copyright 2020 The Fizzy Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <fizzy/execute.hpp>
+#include <fizzy/parser.hpp>
+#include <test/utils/floating_point_utils.hpp>
+#include <test/utils/hex.hpp>
+#include <cfenv>
+#include <iomanip>
+#include <iostream>
+#include <stdexcept>
+#include <string>
+#include <unordered_map>
+
+using namespace fizzy::test;
+
+namespace
+{
+constexpr auto USAGE = R"(Fizzy TestFloat
+
+Tool for testing IEEE 754 floating-point compliance.
+Expects inputs from TestFloat's testfloat_gen.
+See http://www.jhauser.us/arithmetic/TestFloat-3/doc/TestFloat-general.html.
+
+Usage:
+  fizzy-testfloat [options] <function>
+
+  -rnear_even     Round to nearest/even [default].
+  -rminMag        Round to minimum magnitude (toward zero).
+  -rmin           Round to minimum (down).
+  -rmax           Round to maximum (up).
+
+
+Example:
+  testfloat_gen f32_add | fizzy-testfloat f32_add
+)";
+
+// TODO: Implement <float>_roundToInteger functions, issues with them has been spotted.
+
+// The maximum number of failures reported.
+constexpr auto max_failures = 5;
+
+enum class Type
+{
+    i32,
+    i64,
+    f32,
+    f64
+};
+
+enum class Options
+{
+    None = 0,
+
+    /// The Wasm trap happens iff IEEE "invalid operation" exception.
+    TrapIsInvalidOperation = 1,
+};
+
+enum FPException : uint8_t
+{
+    /// An operand is invalid for the operation about to be performed.
+    InvalidOperation = 0x10,
+};
+
+struct FunctionDescription
+{
+    const fizzy::FuncIdx idx;
+    const Type result_type;
+    const Type param_types[2];
+    const size_t num_arguments;
+    const Options options = Options::None;
+
+    constexpr FunctionDescription(fizzy::FuncIdx func_idx, Type result, Type param) noexcept
+      : idx{func_idx}, result_type{result}, param_types{param, Type::f32}, num_arguments{1}
+    {}
+
+    constexpr FunctionDescription(
+        fizzy::FuncIdx func_idx, Type result_ty, Type param1, Type param2) noexcept
+      : idx{func_idx}, result_type{result_ty}, param_types{param1, param2}, num_arguments{2}
+    {}
+
+    constexpr FunctionDescription(
+        fizzy::FuncIdx func_idx, Type result, Type param, Options opts) noexcept
+      : idx{func_idx},
+        result_type{result},
+        param_types{param, Type::f32},
+        num_arguments{1},
+        options{opts}
+    {}
+};
+
+auto get_wasm_code()
```

Comment:
Oh, almost like JS at this point.

Reply:
The only point of this function is to have wast code just before the `FunctionDescription` map.